### DISCUSSION TITLE
Timezones a-la chrono-tz

### DISCUSSION THREAD

**Discussion by jorgecarleitao** [time-rs/time#351] [Question] [answered]
Fixed offsets do not take summer times and others into account. Is there a sibling crate with definitions of other timezones?

We use `chrono` in [arrow2](https://github.com/jorgecarleitao/arrow2) but I am a bit wary of it being unmaintained. However, we currently depend on chrono and chrono-tz for more useful timezones.

Operations where we use chrono and chrono-tz:

```rust
/// Adds an `interval` to a `timestamp` in `time_unit` units without timezone.
pub fn add_naive_interval(timestamp: i64, time_unit: TimeUnit, interval: months_days_ns) -> i64;

/// Adds an `interval` to a `timestamp` in `time_unit` units and timezone `timezone`.
pub fn add_interval<T: chrono::TimeZone>(
    timestamp: i64,
    time_unit: TimeUnit,
    interval: months_days_ns, // [i32, i32, i64]
    timezone: &T,
) -> i64

/// Time units defined in Arrow.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum TimeUnit {
    /// Time in seconds.
    Second,
    /// Time in milliseconds.
    Millisecond,
    /// Time in microseconds.
    Microsecond,
    /// Time in nanoseconds.
    Nanosecond,
}

/// The in-memory representation of the MonthDayNano variant of the "Interval" logical type.
pub struct months_days_ns(i32, i32, i64);
```

Adding days and ns to a datetime with a timezone requires these non-trivial time-shifts.

**Comment by jhpratt** [comment, marked as answer]
Looks like you're asking for two separate things here: tzdb integration and a `Period` type. Both are planned.

tzdb support (tracked by #193) is what will provide time zones rather than just fixed offsets. This will let you subtract two dates over a DST shift and get the appropriate `Duration` (such as 23 or 25 hours rather than 24).

A `Period` type is what will allow you to take January 1 and add one month, rather than requiring a number of days (which of course varies by month). It would be able to be added to and subtracted from all of the types you'd expect. That will include `ZonedDateTime` or whatever it ends up being called.

**Reply by jorgecarleitao** [reply]
Thanks! Exactly.

fwiw, the period is not so relevant because getting leap days is relatively easy with a `Date` and duration:

```rust
fn add_month(year: i32, month: u32, months: i32) -> chrono::NaiveDate {
    let new_year = (year * 12 + (month - 1) as i32 + months) / 12;
    let new_month = (year * 12 + (month - 1) as i32 + months) % 12 + 1;
    chrono::NaiveDate::from_ymd(new_year, new_month as u32, 1)
}

fn get_days_between_months(year: i32, month: u32, months: i32) -> i64 {
    add_month(year, month, months)
        .signed_duration_since(chrono::NaiveDate::from_ymd(year, month, 1))
        .num_days()
}
```

Since datetimes in Apache Arrow are stored as the equivalent of `(Vec<i64>, Option<String>)`, we do not need a `ZonedDateTime` (although I do see the need for other users ^_^)

**Reply by jhpratt** [reply]
Sure, but at that point you already have the end. May as well just calculate it directly, which is what a `Period` type would do.

In the future you will be able to do something like

```rust
datetime!(2021-01-01 0:00 America/New_York) + Period { months: 2, days: 6, seconds: 18, ..Period::ZERO }
```

and get 2021-03-07 0:00:18 in US eastern time. If performed on the same date in 2020, you'd get the same result (obviously in 2020). It looks like that is what you're looking for from the `TimeUnit` enum, no?

**Reply by jorgecarleitao** [reply]
The end is adding `months_days_ns` to a zonedDateTime just like you wrote. I was just pointing out that the `tzdb` is necessary to work with periods, while periods without tzdb only work up to days for zoned datetimes.

I think we are in agreement here :)

**Reply by jhpratt** [reply]
Nah, you could definitely do `datetime!(2021-01-01 0:00 -5:00) + Period { months: 2, days: 6, seconds: 18, ..Period::ZERO }`. For reasons that you appear to understand, that wouldn't change the UTC offset if it crosses a DST boundary. But it _is_ possible to implement on things other than a `ZonedDateTime` which wouldn't require tzdb.

**Reply by jorgecarleitao** [reply]
ahh, I see, you are right, they are independent indeed. Thanks for the clarification.

**Reply by jhpratt** [reply]
Once I (eventually) get around to cleaning up docs and writing at least part of the book for time 0.3, a `Period` type is next in priority.
